raco/rules.py

```diff
@@ -406,8 +406,11 @@ def descend_tree(op, cond):
     def fire(self, op):
         if not isinstance(op, algebra.Select):
             return op
-        if op.has_been_pushed:
-            return op
+        if hasattr(op, "has_been_pushed"):
+            if op.has_been_pushed:
+                return op
+            else:
+                 op.has_been_pushed = False
```

Comment:
Indeed, I screwed up the indentation when fixing the travis formatting
checks.

The else: should be associated with the hasattr check: If the attribute
isn't present, add it and make it false.

On Mon, Apr 6, 2015 at 9:27 AM, Brandon Myers <notifications@github.com>
wrote:

> In raco/rules.py
> <https://github.com/uwescience/raco/pull/404#discussion_r27808050>:
>
> > @@ -406,8 +406,11 @@ def descend_tree(op, cond):
> >      def fire(self, op):
> >          if not isinstance(op, algebra.Select):
> >              return op
> > -        if op.has_been_pushed:
> > -            return op
> > +        if hasattr(op, "has_been_pushed"):
> > +            if op.has_been_pushed:
> > +                return op
> > +            else:
> > +                 op.has_been_pushed = False
>
> This seems redundant. If has_been_pushed is False, make it False. Or is it
> that has_been_pushed is 0 or None somewhere?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/uwescience/raco/pull/404/files#r27808050>.
>



-- 
Bill Howe
Associate Director and Senior Data Science Fellow, UW eScience Institute
Affiliate Faculty, Computer Science & Engineering
University of Washington
To acknowledge eScience: "Supported in part by the University of Washington
eScience Institute"